File: 29_無限ループ/python/main.py
#!/usr/bin/env python
# coding: -utf8

import user_input as ui


def validate_rate(s):
    rate = int(s)
    if rate == 0:
        raise ValueError
    return rate


if __name__ == '__main__':
    rate = ui.read_user_input_with_validator(
        "What is the rate of return? ", validate_rate,  "Sorry. That's not a valid input.")

    years = int(72 / rate)
    print(
        f"It will take {years} years to double your initial investment.")

File: 29_無限ループ/go/main.go
package main

import (
	"fmt"
	"strconv"
)

func main() {
	rate := readUserInputWithValidator(
		"What is the rate of return? ",
		validateRate,
		"Sorry. That's not a valid input.")
	years := 72 / rate
	fmt.Printf("It will take %d years to double your initial investment.\n", years)
}

func validateRate(input string) (bool, int) {
	rate, err := strconv.Atoi(input)
	if err != nil {
		return false, 0
	}
	if rate == 0 {
		return false, 0
	}
	return true, rate
}
